apps/hash-graph/lib/graph/src/snapshot/ontology/property_type/batch.rs

```diff
@@ -0,0 +1,144 @@
+                        (LIKE property_types INCLUDING ALL)
+                        ON COMMIT DROP;
+
+                    CREATE TEMPORARY TABLE property_type_constrains_values_on_tmp (
+                        source_property_type_ontology_id UUID NOT NULL,
+                        target_data_type_base_url TEXT NOT NULL,
+                        target_data_type_version INT8 NOT NULL
+                    ) ON COMMIT DROP;
+
+                    CREATE TEMPORARY TABLE property_type_constrains_properties_on_tmp (
+                        source_property_type_ontology_id UUID NOT NULL,
+                        target_property_type_base_url TEXT NOT NULL,
+                        target_property_type_version INT8 NOT NULL
+                    ) ON COMMIT DROP;
+                ",
+            )
+            .await
+            .into_report()
+            .change_context(InsertionError)
+            .attach_printable("could not create temporary tables")?;
+        Ok(())
+    }
+
+    async fn write(&self, postgres_client: &PostgresStore<C>) -> Result<(), InsertionError> {
+        let client = postgres_client.as_client().client();
+        match self {
+            Self::Schema(property_types) => {
+                let rows = client
+                    .query(
+                        r"
+                            INSERT INTO property_types_tmp
+                            SELECT DISTINCT * FROM UNNEST($1::property_types[])
+                            RETURNING 1;
+                        ",
+                        &[property_types],
+                    )
+                    .await
+                    .into_report()
+                    .change_context(InsertionError)?;
+                if !rows.is_empty() {
+                    tracing::info!("Read {} property type schemas", rows.len());
+                }
+            }
+            Self::ConstrainsValues(values) => {
+                let rows = client
+                    .query(
+                        r"
+                            INSERT INTO property_type_constrains_values_on_tmp
+                            SELECT DISTINCT * FROM UNNEST($1::property_type_constrains_values_on_tmp[])
+                            RETURNING 1;
+                        ",
+                        &[values],
+                    )
+                    .await
+                    .into_report()
+                    .change_context(InsertionError)?;
+                if !rows.is_empty() {
+                    tracing::info!("Read {} property type value constrains", rows.len());
+                }
+            }
+            Self::ConstrainsProperties(properties) => {
+                let rows = client
+                    .query(
+                        r"
+                            INSERT INTO property_type_constrains_properties_on_tmp
+                            SELECT DISTINCT * FROM UNNEST($1::property_type_constrains_properties_on_tmp[])
```

Comment:
How does this casting work? Is it the name of the table we cast as and `to_sql` knows how to convert records into the expected shape?

Reply:
Very cool, good to know!